### ISSUE TITLE
User-defined type guard behaves differently inside `Array.filter`

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#30240]
**TypeScript Version:**  3.3.3333

**Search Terms:** user defined type guard array filter narrow

**Code**

```ts
type BasicUserData = { name: string };
type UserStats = { stats: {} };

type User = BasicUserData | UserStats | (BasicUserData & UserStats);

declare const checkHasUserStats: (user: User) => user is UserStats;

declare const user: BasicUserData | (BasicUserData & UserStats);

// no type error, good
user.name;
// type error, good
user.stats;

if (checkHasUserStats(user)) {
    // Expected and actual `typeof user`: BasicUserData & UserStats

    // no type error, good
    user.name;
    // no type error, good
    user.stats;
}

declare const users: Array<BasicUserData | (BasicUserData & UserStats)>;

users.filter(checkHasUserStats).map(user => {
    // Expected `typeof user`: BasicUserData & UserStats
    // Actual: BasicUserData

    // no type error, good
    user.name;
    // unexpected type error, bad (!)
    user.stats;
});
```

**Comment by jack-williams** [comment]
Related: #29501

**Comment by RyanCavanaugh** [comment]
The problem is here: `BasicUserData | (BasicUserData & UserStats)` - this is a subtype-reducing union that is going to evaporate if poked.

Declaring this as
```ts
declare const users: Array<BasicUserData & Partial<UserStats>>;
```
causes the expected behavior.

**Comment by OliverJAsh** [comment]
> this is a subtype-reducing union that is going to evaporate if poked.

Can you elaborate on why? It seems there's a caveat in the type system here that I need to be aware of, so any information you can share would be appreciated!

Also, I'm still not sure I understand why this behaviour isn't consistent between `filter` and `if` usages of the type guard.

> ```ts
> Array<BasicUserData & Partial<UserStats>>;
> ```

IIUC, those types aren't entirely accurate, as `Partial` allows any of the properties in `UserStats` to be undefined, whereas I want to say "either no `A` or all of `A`".

In my contrived example this doesn't matter much because `UserStats` has a single property, but that might not always be true. (Indeed it isn't in my real world example where I initially discovered this issue.)

**Comment by jack-williams** [comment]
The issue here is that your filter predicate is not a valid type guard when instantiated. The refinement of a predicate needs to be a assignable to the input. You can see the issue here:
```ts
type BasicUserData = { name: string };
type UserStats = { stats: {} };
declare const users: Array<BasicUserData | (BasicUserData & UserStats)>;
declare const checkHasUserStats: (user: typeof users[number]) => user is UserStats; // error
```
The type guard does not typecheck if it's written as it would be used in the filter.

TLDR: `UserStats` is not assignable to `BasicUserData | (BasicUserData & UserStats)`, so the type guard overload is not selected by `filter`.

Using @RyanCavanaugh's solution the type guard overload is still not selected, but the type `BasicUserData & Partial<UserStats>` is more permissive an lets you assign to property `stats`.

**Comment by OliverJAsh** [comment]
@jack-williams That explains a lot I think, thank you.

One question I have

> filter predicate is not a valid type guard when instantiated

I instantiate the type guard with exactly the same type when using `if`, and it seems to work OK there—so I'm still not entirely clear why this behaviour isn't consistent between `if` and `filter`. 🤔

**Comment by jack-williams** [comment]
The difference (I believe) is that in the `if` you deal with the concrete instantiation of the type guard immediately.

In the filter case I think the issue is due to inference. Basically `T` has multiple inference candidates from contravariant positions (`BasicUserData | (BasicUserData & UserStats`), and covariant positions (`User`). The inference seems to favour contravariant types unless the covariant ones are more precise, which is not the case here.

There is no reasoning that understands that picking the weaker inference for `T` would enable the typeguard to be active, which is probably what the users wants.

**Comment by OliverJAsh** [comment]
@jack-williams Do you think any suggestions or bugs need to be filed as a result of what you're describing? I'm certain the behaviour I described is going to confuse users, especially when the behaviour differs between `if` and `filter`.

**Comment by OliverJAsh** [comment]
@jack-williams 

> Writing your typeguard like this will give the desired behaviour I think.
> 
> ```ts
> declare const checkHasUserStats: <T extends User>(user: T) => user is T & UserStats;
> ```

That achieves the desired behaviour, however it creates another problem: if we use a constrained generic then it seems we lose the ability to validate the type guard:

```ts
const checkHasUserStats = (user: User): user is UserStats => {
    if ('stats' in user) {
        // No error, good! :-)
        const _test: UserStats = user;
        return true;
    } else {
        return false;
    }
};
const checkHasUserStatsGeneric = <T extends User>(user: T): user is T & UserStats => {
    if ('stats' in user) {
        // Error, bad! :-(
        const _test: UserStats = user;
        return true;
    } else {
        return false;
    }
};
```

Do you have any suggestions for that?
